Suggestions for the migration docs (#2718)
apollographql/apollo-ios: docs/source/migrations/1.0.mdx

import SetupCodegen from "../../shared/setup-codegen/combined.mdx"
import SeparateLocalCacheMutationsNote from "../../shared/separate-local-cache-mutation-note.mdx"

Apollo iOS 1.0 provides a stable API that uses modern Swift language conventions and features.

Among other improvements, Apollo iOS 1.0 features new:

- Code generation tooling written in pure Swift code.
- Generated models that improve readability and functionality while reducing generated code size.
- Support for using generated models in multi-module projects.
- Type-safe APIs for cache key resolution.

This article describes the significant changes for this major version of Apollo iOS and a walk-through of migrating an existing project from Apollo iOS 0.x to 1.0.

> **Steps to migrate to Apollo iOS 1.0:**
> 1. Begin by reading about the [key changes](#key-changes).

> > 2. Next, follow our [step-by-step instructions](#step-by-step-instructions) to upgrade your existing app.

Reviewer: I don't know why we wanted to change the heading from "Step-by-step instructions" to "Migration walk-through", but we need to then change this link also.

I personally prefer step-by-step instructions, but am fine with either, as long as we fix the link!

> 3. Finally, see [breaking changes](#breaking-changes) to resolve any remaining build errors.

## Key changes

### Generated schema module

The 0.x version of Apollo iOS generates models for your GraphQL operation definitions _and_ the input objects and enums referenced in your schema.

Apollo iOS 1.0 expands on this, generating an _entire module_ that contains models and metadata for your GraphQL schema and its type definitions.

In addition to your input objects and enum types, this module contains:
- Types for the objects, interfaces, and unions in your schema.
- Editable definitions for your custom scalars.
- An editable `SchemaConfiguration.swift` file.
- Metadata for the Apollo GraphQL executor.

> Schema types are contained in their own namespace to prevent naming conflicts. You can generate this namespace as a stand-alone module, imported by your project, or as a caseless namespace enum that you can embed in your target application.

Reviewer: I think the end of this should still be "application target" not "target application".

Reviewer 2: Yes, agreed. There is a subtle difference.

> Schema types are contained in their own namespace to prevent naming conflicts. You can generate this namespace as a stand-alone module, imported by your project, or as a caseless namespace enum that you can embed in your target application.

Reviewer: ```suggestion
Schema types are contained in their own namespace to prevent naming conflicts. You can generate this namespace as a stand-alone module, imported by your project, or as a caseless namespace enum that you can embed in your application target.
```

### Multi-module support

> Apollo iOS 1.0 can support complex applications composed of multiple modules _and_ monolithic application targets.

Reviewer: Not "and", this should be "or"

> Apollo iOS 1.0 can support complex applications composed of multiple modules _and_ monolithic application targets.

Reviewer: ```suggestion
Apollo iOS 1.0 can support complex applications composed of multiple modules _or_ monolithic application targets.
```

The [generated schema module](#generated-schema-module) supports multi-module projects by containing all of a schema's shared types and metadata. This enables you to move generated operation models anywhere within your project structure (as long as they are linked to the schema module).

Apollo iOS's code generation engine provides flexible configuration options that make code generation work seamlessly for any project structure.

> ## Migration walk-through

Reviewer: ```suggestion
## Step-by-step instructions
```

> > **Before migrating to Apollo iOS 1.0, consider your project structure and how you want to include your generated schema module and operation models.**

Reviewer: ```suggestion
> **Before migrating to Apollo iOS 1.0, you should consider your project structure and decide how you want to include your generated schema module and operation models.**
```

> > **Before migrating to Apollo iOS 1.0, consider your project structure and how you want to include your generated schema module and operation models.**

Reviewer: I don't like this new wording. The old wording may be a little verbose, but it is actually pretty important that people do this, and I feel like this sentence doesn't make it clear how imperative it is that people take a minute to consider their project's organization before starting on the migration.

Not understanding how you want to structure your project (and configuring code generation accordingly) has been the #1 cause of confusion (and GitHub issues) since 1.0 released.

>

> > [Learn how to configure Apollo iOS for your project structure](../project-configuration).

Reviewer: ```suggestion
> To learn about how you can best integrate Apollo iOS to suit your project's needs, see our [Project configuration documentation](./project-configuration).
```

To migrate to Apollo iOS 1.0, you'll do the following:
1. [Update your Apollo iOS dependency](#1-update-to-apollo-ios-10)

> 2. [Setup code generation](#2-setup-code-generation)

Reviewer: ```suggestion
2. [Setup code generation](#2-set-up-code-generation)
```

Then, update the `ApolloCodegenConfiguration` in your script with the new configuration values. For a list of configuration options, see [Codegen configuration](./../code-generation/codegen-configuration).

### Step 3: Replace the code generation build phase

> We no longer recommend running Apollo's code generation as an Xcode build phase.

Your generated files change whenever you modify your `.graphql` operation definitions (which happens infrequently). Running code generation on every build increases build times and slows development. 

Instead, we recommend running code generation **manually (using the CLI)** whenever you modify your `.graphql` files.

If you want to continue running code generation on each build, you can update your build script to run the CLI `generate` command.

### Step 4: Refactor your code

While designing Apollo iOS 1.0, we tried to limit the number of code changes required to migrate from legacy versions.

Below are explanations for each [breaking change](#breaking-changes) that Apollo iOS 1.0 brings and tips on addressing those changes during your migration.

## Breaking changes

### Custom scalars

> > If you _didn't_ define any custom scalar types or use the `--passthroughCustomScalars` code generation option, you can skip this step. 

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion
```
